### PR TITLE
[SM-1293] Add endpoint to fetch secret's access policies

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-1293

## 📔 Objective

The objective of this PR is to add the ability to fetch a secret's access policies.
This change includes consolidating access policy response models so they can be used for secret access policies.

clients PR: https://github.com/bitwarden/clients/pull/9463

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: src/Api/SecretsManager/Models/Response/AccessPolicyResponseModel.cs [hunk 1]
```diff
@@ -9,161 +9,133 @@ public abstract class BaseAccessPolicyResponseModel : ResponseModel
...
+    public string? ServiceAccountName { get; set; }
 }
 
-public class ServiceAccountProjectAccessPolicyResponseModel : BaseAccessPolicyResponseModel
+public class GrantedProjectAccessPolicyResponseModel : BaseAccessPolicyResponseModel
 {
     private const string _objectName = "serviceAccountProjectAccessPolicy";
 
-    public ServiceAccountProjectAccessPolicyResponseModel(ServiceAccountProjectAccessPolicy accessPolicy)
+    public GrantedProjectAccessPolicyResponseModel(ServiceAccountProjectAccessPolicy accessPolicy)
```
Comment (mzieniukbw) [inline src/Api/SecretsManager/Models/Response/AccessPolicyResponseModel.cs:127]: update `_objectName`

File: src/Api/SecretsManager/Models/Response/GrantedProjectAccessPolicyPermissionDetailsResponseModel.cs [hunk 2]
```diff
@@ -4,22 +4,22 @@
 
 namespace Bit.Api.SecretsManager.Models.Response;
 
-public class ServiceAccountProjectAccessPolicyPermissionDetailsResponseModel : ResponseModel
+public class GrantedProjectAccessPolicyPermissionDetailsResponseModel : ResponseModel
 {
     private const string _objectName = "serviceAccountProjectAccessPolicyPermissionDetails";
 
-    public ServiceAccountProjectAccessPolicyPermissionDetailsResponseModel(
+    public GrantedProjectAccessPolicyPermissionDetailsResponseModel(
```
Comment (mzieniukbw) [inline src/Api/SecretsManager/Models/Response/GrantedProjectAccessPolicyPermissionDetailsResponseModel.cs:11]: update `_objectName`

**Review by mzieniukbw — CHANGES_REQUESTED** [review]

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs [hunk 3]
```diff
@@ -47,6 +47,9 @@ public class SecretAuthorizationHandler : AuthorizationHandler<SecretOperationRe
             case not null when requirement == SecretOperations.Delete:
                 await CanDeleteSecretAsync(context, requirement, resource);
                 break;
+            case not null when requirement == SecretOperations.ReadAccessPolicies:
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:50]: This is resource based authorization https://learn.microsoft.com/en-us/aspnet/core/security/authorization/resourcebased?view=aspnetcore-8.0. 

We want to ask does the user have access to read this secret's (resource's) access policies.

Creating a new handler at  `bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/AccessPolicies` would require duplicating the `SecretAuthorizationHandler` again or basing it on another resource.

File: bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/AccessPolicyRepository.cs [hunk 4]
```diff
@@ -395,6 +416,42 @@ await dbContext.ServiceAccountProjectAccessPolicy
...
 
+    public async Task<SecretAccessPolicies?> GetSecretAccessPoliciesAsync(
+        Guid secretId,
+        Guid userId)
+    {
+        await using var scope = ServiceScopeFactory.CreateAsyncScope();
+        var dbContext = GetDatabaseContext(scope);
+
+        var entities = await dbContext.AccessPolicies.Where(ap =>
+                ((UserSecretAccessPolicy)ap).GrantedSecretId == secretId ||
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/AccessPolicyRepository.cs:427]: In normal C# that would be a concern. 

This is an Entity Framework Core query, so it's different. 

It takes the C# code and translates it into a SQL query. That SQL query is what actually gets executed.

The `AccessPolicy` table is set up as a [Table-per-hierarchy](https://learn.microsoft.com/en-us/ef/core/modeling/inheritance#table-per-hierarchy-and-discriminator-configuration) configuration that is why we can do that castings here.

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs [hunk 5]
```diff
@@ -152,6 +155,26 @@ public class SecretAuthorizationHandler : AuthorizationHandler<SecretOperationRe
...
+
+        // Only users and admins can read access policies
+        if (accessClient != AccessClientType.User && accessClient != AccessClientType.NoAccessCheck)
+        {
+            return;
+        }
+
+        var access = await _secretRepository.AccessToSecretAsync(resource.Id, userId, accessClient);
+
+        if (access.Write)
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:171]: Correct if you have read-only access, you can only view the secret.

Project access is the same today. With read-only access, you can view the project but not the project's access policies.

File: src/Api/SecretsManager/Models/Response/AccessPolicyResponseModel.cs (hunk 1 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Models/Response/AccessPolicyResponseModel.cs:127]: Great catch, updated in https://github.com/bitwarden/server/pull/4146/commits/cc7f5dd85e2e29873417c53e5c134685c0f27aa3

File: src/Api/SecretsManager/Models/Response/GrantedProjectAccessPolicyPermissionDetailsResponseModel.cs (hunk 2 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Models/Response/GrantedProjectAccessPolicyPermissionDetailsResponseModel.cs:11]: Great catch, updated in https://github.com/bitwarden/server/pull/4146/commits/cc7f5dd85e2e29873417c53e5c134685c0f27aa3

**Review by mzieniukbw — APPROVED** [review]

**Pull request merged**